### ISSUE TITLE
test/buffer002 fails on x86

### ISSUE THREAD

**Issue by jacereda** [idris-lang/Idris-dev#4400]
Since ints are unboxed and the type used is intptr_t, there's a different result on 32-bit boxes due to sign propagation.

Shouldn't Data.Buffer use Bits32 instead?

```
    buffer002:         FAIL (14.17s)
      --- test/buffer002/output	2018-03-30 22:04:14.218527462 +0200
      +++ test/buffer002/expected	2018-03-20 09:11:17.843891452 +0100
      @@ -1,9 +1,9 @@
       40
      -[00, 00, 00, 00, 00, D2, 02, 96, C9, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 48, 65, 6C, 6C, 6F, 20, 77, 6F, 72, 6C, 64, 21, 00, 00, 00, 00, D2, 02, 96, C9]
      --912915758
      -13211138
      +[00, 00, 00, 00, 00, D2, 02, 96, 49, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 48, 65, 6C, 6C, 6F, 20, 77, 6F, 72, 6C, 64, 21, 00, 00, 00, 00, D2, 02, 96, 49]
      +1234567890
      +4822530
       123.456
      --912915758
      +1234567890
       0
       Hello worl
       Hello world!
```

**Comment by melted** [comment]
It might have Bits32 too, but it should be easy to use and needs to support negative numbers.

The result on 32-bit is just wrong though, there is no sign that should be propagated. I wonder where that comes from.

**Comment by melted** [comment]
Oh yeah, the int tag makes ints on 32-bit effectively 31-bit, and 1234567890 is high enough to be a negative 31-bit number.

if I change the value written to 1034567890, it gives the right result.

**Comment by jacereda** [comment]
Yes, so I was wondering if it's desirable to have different behaviour on 32-bit platforms for this library. If only Bits values were allowed it would be more consistent across platforms.

**Comment by melted** [comment]
I think it's OK to add setBit32. But we should keep setInt, because it will actually work for all valid values of Int on 32-bit systems. 

The bug here is that the 1234567890 literal is accepted as an Int on 32-bit Idris. It can't work, because it's indistinguishable from -912915758.

**Comment by jfdm** [comment]
@melted Was this supposed to be closed by #4406? I think the title didn't follow github's conventions for auto closing tickets fixed by PRs.

**Comment by melted** [comment]
Ah, thanks.
